src/components/App/index.tsx

```diff
@@ -55,12 +54,9 @@ export class AppBase extends React.Component<Props> {
     const { apiState, dispatch, profile } = this.props;
 
     if (!profile && prevProps.apiState.authToken !== apiState.authToken) {
-      const response = (await api.callApi({
-        apiState,
-        endpoint: '/accounts/profile/',
-      })) as ExternalUser;
+      const response = await getCurrentUserProfile(apiState);
 
-      if (response && response.name) {
+      if (!isErrorResponse(response)) {
```

Comment:
This seems like the wrong place to defend against errors. It's awkward and adds unnecessary complexity. I know your code is no worse than the code it replaces but I would rather see something like this:

```diff
diff --git a/src/api/index.spec.tsx b/src/api/index.spec.tsx
index 3597be2..69c95d2 100644
--- a/src/api/index.spec.tsx
+++ b/src/api/index.spec.tsx
@@ -210,6 +210,8 @@ describe(__filename, () => {
 
   describe('getCurrentUserProfile', () => {
     it('calls the API to retrieve the current logged-in user profile', async () => {
+      // Something like this needs to get applied everywhere.
+      fetchMock.mockResponse(JSON.stringify({}));
       await getCurrentUserProfile(defaultApiState);
 
       expect(fetch).toHaveBeenCalledWith(`/api/v4/accounts/profile/`, {
diff --git a/src/api/index.tsx b/src/api/index.tsx
index a75f91d..02a8e21 100644
--- a/src/api/index.tsx
+++ b/src/api/index.tsx
@@ -71,6 +71,7 @@ export const callApi = async <SuccessResponseType extends {}>({
     adjustedEndpoint = `${adjustedEndpoint}?${queryString}`;
   }
 
+  let responseData;
   try {
     const response = await fetch(`/api/${version}${adjustedEndpoint}`, {
       method,
@@ -85,14 +86,19 @@ export const callApi = async <SuccessResponseType extends {}>({
       );
     }
 
-    return await response.json();
+    responseData = await response.json();
   } catch (error) {
-    log.debug('Error caught in callApi():', error);
-
-    return {
+    responseData = {
       error,
     };
   }
+
+  if (isErrorResponse(responseData)) {
+    log.debug('Error returned from callApi():', responseData.error);
+    throw new Error(`callApi() error: ${responseData.error}`);
+  }
+
+  return responseData;
 };
 
 type GetVersionFileParams = {
```

There might be cases where the code executing `callApi()` will expect an error object but I don't think we have that code yet. When we need it, we can add a parameter like `expectError: true` or something.

Reply:
I still believe we should not `throw` in `callApi`, but this depends on how we plan to handle errors. I guess #182 just got a higher priority now :)